flow/ProtocolVersion.h

```diff
@@ -0,0 +1,123 @@
+template <class P>
+using latest_version = v6_1<P>;
+
+} // namespace fdb_versions
+
+// ProtocolVersion wraps a uint64_t to make it type safe. It has to inherit
+// from the newest version object and it will know about the current version.
+// The default constuctor will initialize the version to 0 (which is an invalid
+// version). ProtocolVersion objects should never be compared to version numbers
+// directly. Instead one should always use the type-safe version types from which
+// this class inherits all.
+class ProtocolVersion : public fdb_versions::latest_version<ProtocolVersion> {
+	uint64_t _version;
+public: // constants
+	static constexpr uint64_t versionFlagMask = 0x0FFFFFFFFFFFFFFFLL;
+	static constexpr uint64_t objectSerializerFlag = 0x1000000000000000LL;
+	static constexpr uint64_t compatibleProtocolVersionMask = 0xffffffffffff0000LL;
+	static constexpr uint64_t minValidProtocolVersion = 0x0FDB00A200060001LL;
+
+public:
+	constexpr explicit ProtocolVersion(uint64_t version) : _version(version) {}
+	constexpr ProtocolVersion() : _version(0) {}
+
+	constexpr bool isCompatible(ProtocolVersion other) const {
+		return (other.version() & compatibleProtocolVersionMask) == (version() & compatibleProtocolVersionMask);
+	}
+	constexpr bool isValid() const { return version() >= minValidProtocolVersion; }
+
+	constexpr uint64_t version() const { return _version & versionFlagMask; }
+	constexpr uint64_t versionWithFlags() const { return _version; }
+
+	constexpr bool hasObjectSerializerFlag() const { return (_version & objectSerializerFlag) > 0; }
+	constexpr void addObjectSerializerFlag() { _version = _version | objectSerializerFlag; }
+	constexpr void removeObjectSerializerFlag() {
+		_version = hasObjectSerializerFlag() ? _version ^ objectSerializerFlag : _version;
+	}
+	constexpr void removeAllFlags() { _version = version(); }
+
+	constexpr operator bool() const { return _version != 0; }
+	// comparison operators
+	constexpr bool operator==(const ProtocolVersion other) const { return version() == other.version(); }
+	constexpr bool operator!=(const ProtocolVersion other) const { return version() != other.version(); }
+	constexpr bool operator<=(const ProtocolVersion other) const { return version() <= other.version(); }
+	constexpr bool operator>=(const ProtocolVersion other) const { return version() >= other.version(); }
+	constexpr bool operator<(const ProtocolVersion other) const { return version() < other.version(); }
+	constexpr bool operator>(const ProtocolVersion other) const { return version() > other.version(); }
+};
+
+// These impact both communications and the deserialization of certain database and IKeyValueStore keys.
+//
+// The convention is that 'x' and 'y' should match the major and minor version of the software, and 'z' should be 0.
+// To make a change without a corresponding increase to the x.y version, increment the 'dev' digit.
+//
+//                                                        xyzdev
```

Comment:
It looks like these arrows are pointing one digit to the left of where they should be

Reply:
you're right, I fixed this.